ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs

```diff
@@ -226,127 +244,162 @@ initLedgerDB replayTracer
-        let tracer' = decorateReplayTracerWithStart initialPoint tracer
-        (initDB, replayed) <-
-          initStartingWith
-            tracer'
-            cfg
-            backingStore
-            streamAPI
-            (ledgerDbWithAnchor runAlsoLegacy extLedgerSt)
-        return (tip, initDB, replayed, backingStore)
+initFromSnapshotOrGenesis replayTracer snapshotTracer getGenesisLedger hasFS decLedger decHash runAlsoLegacy = do
+    listSnapshots hasFS >>= tryNewestFirst id
+  where
+    tryNewestFirst :: (InitLog blk -> InitLog blk)
+                   -> [DiskSnapshot]
+                   -> m ( InitLog   blk
+                        , LedgerDB' blk
+                        , LedgerBackingStore' m blk
+                        , ReplayTracer m blk
+                        )
+    tryNewestFirst acc [] = do
+      -- We're out of snapshots. Start at genesis
+      traceWith replayTracer ReplayFromGenesis
+      genesisLedger <- getGenesisLedger
+      let replayTracer' = decorateReplayTracerWithStart (Point Origin) replayTracer
+          initDb        = ledgerDbWithAnchor runAlsoLegacy (stowLedgerTables genesisLedger)
+      backingStore <- newBackingStore hasFS (projectLedgerTables genesisLedger) -- TODO: needs to go into ResourceRegistry
+      return ( acc InitFromGenesis
+             , initDb
+             , backingStore
+             , replayTracer'
+             )
+
+    tryNewestFirst acc (s:ss) = do
+      eExtLedgerSt <- runExceptT $ readSnapshot hasFS decLedger decHash s
+      case eExtLedgerSt of
+        Left err -> do
+          when (diskSnapshotIsTemporary s) $
+            deleteSnapshot hasFS s
+          traceWith snapshotTracer . InvalidSnapshot s . InitFailureRead $ err
+          tryNewestFirst (acc . InitFailure s (InitFailureRead err)) ss
+        Right extLedgerSt -> do
+          let initialPoint =
+                withOrigin (Point Origin) annTipPoint
+                $ headerStateTip
+                $ headerState
+                $ extLedgerSt
+          case pointToWithOriginRealPoint (castPoint (getTip extLedgerSt)) of
+            Origin        -> do
+              -- Delete the snapshot of the Genesis ledger state. It should have
+              -- never existed.
+              deleteSnapshot hasFS s
+              traceWith snapshotTracer . InvalidSnapshot s $ InitFailureGenesis
+              tryNewestFirst (acc . InitFailure s InitFailureGenesis) []
+            NotOrigin tip -> do
+              backingStore <- restoreBackingStore hasFS s -- TODO this needs to go in the resource registry
+              traceWith replayTracer $
+                ReplayFromSnapshot s tip (ReplayStart initialPoint)
+              let tracer' = decorateReplayTracerWithStart initialPoint replayTracer
+                  initDb  = ledgerDbWithAnchor runAlsoLegacy extLedgerSt
+                  -- TODO: @js if the provided extLedgerSt has an empty UTxO, then we cannot run the legacy db
```

Comment:
I think your `isCandidateForUnstow` method already addresses this TODO. Did you have more in mind?

Reply:
Indeed, wrote the comment, then wrote the function and forgot to delete the comment. Good catch.